README.md

```diff
@@ -34,17 +34,382 @@ The Flow API provides functionalities for building asynchronous data processing
+     }).retryWith((cause) =>  RetryPolicy.randomisedBackoff())
+       .collect(print);
+   ```
+  
+ ### NoRetryPolicy
+
+ As the name suggests, the NoRetryPolicy implements the retry interface but does not attempt retries. It is useful as a default policy or in situations where retries are not desired.
+
+  ```dart
+     flow((collector) {
+       collector.emit('A');
+       throw Exception('Something went wrong');
+     }).retryWith((cause) =>  RetryPolicy.decorrelatedJitter())
+       .collect(print);
+   ```
+
+### Implementing Custom Retry Policies
+
+To create your custom retry policy, implement the RetryPolicy interface and override the retry method. Your implementation should define the logic to determine if a retry should occur based on the number of [attempts] already made and the nature of the failure.
+
+```dart
+class MyCustomRetryPolicy implements RetryPolicy {
+  @override
+  FutureOr<bool> retry(int attempt) {
+    // Implement your custom retry logic here
+    // Return true to retry, false to not retry
+    if (attempt < 3) {
+      return true; // Retry on specific error up to 3 times
+    }
+    return false; // Do not retry for other errors or after 3 attempts
+  }
+}
+
 ```
 
+
+## onCompletion
+Executes an action upon flow completion (needs improvement).
+
+This function allows you to perform actions or cleanup tasks after the flow has finished processing data. The provided `action` function receives an optional exception (`null` if no exception occurred) and the `FlowCollector` as arguments. **Note:** Currently, the handling of completion errors within the context of the flow needs improvement.
+
+  ```dart
+  flowOf([1, 2, 3]).onCompletion((exception, collector){
+        //Perform  action
+    });
+  ```
+
+[action] : A function that takes an optional `Exception` and a  `FlowCollector<T>` as arguments. It can be used for post-processing, cleanup, or handling any errors that might occur during completion.
+
+## onStart
+Executes an action before the flow starts collecting data.
+
+This function allows you to perform setup tasks or initializations before the actual flow processing begins. The provided `action` function receives the `FlowCollector` as an argument.
+
+ ```dart
+   flow((collector) => collector.emit('World!'))
+     .onStart((collector) => collector.emit('Hello,'));
+     .collect(stdout.write)
+
+   // Outputs:
+   // Hello, World!
+ ```
+
+[action] : A function that takes a `FlowCollector<T>` as an argument. It can be used for pre-processing or any actions needed before collecting data in the flow.
+
+## onEach
+Returns a flow that invokes the given [action] before each value of theupstream flow is emitted downstream.
+  
+This function allows you to perform actions on each individual value that flows through the pipeline, potentially performing side effects before the value is sent further downstream.
+  
+   ```dart
+    flow((collector) => collector.emit(1, 2, 3))
```

Comment:
The example here can be very misleading , the emit function only takes a single argument and not multiple arguments.